internal/concierge/impersonator/impersonator.go

```diff
@@ -0,0 +1,161 @@
+)
+
+// allowedHeaders are the set of HTTP headers that are allowed to be forwarded through the impersonation proxy.
+//nolint: gochecknoglobals
+var allowedHeaders = []string{
+	"Accept",
+	"Accept-Encoding",
+	"User-Agent",
+	"Connection",
+	"Upgrade",
+}
+
+type proxy struct {
+	cache *authncache.Cache
+	proxy *httputil.ReverseProxy
+	log   logr.Logger
+}
+
+func New(cache *authncache.Cache, log logr.Logger) (http.Handler, error) {
+	return newInternal(cache, log, func() (*rest.Config, error) {
+		client, err := kubeclient.New()
+		if err != nil {
+			return nil, err
+		}
+		return client.JSONConfig, nil
+	})
+}
+
+func newInternal(cache *authncache.Cache, log logr.Logger, getConfig func() (*rest.Config, error)) (*proxy, error) {
+	kubeconfig, err := getConfig()
+	if err != nil {
+		return nil, fmt.Errorf("could not get in-cluster config: %w", err)
+	}
+
+	serverURL, err := url.Parse(kubeconfig.Host)
+	if err != nil {
+		return nil, fmt.Errorf("could not parse host URL from in-cluster config: %w", err)
+	}
+
+	kubeTransportConfig, err := kubeconfig.TransportConfig()
+	if err != nil {
+		return nil, fmt.Errorf("could not get in-cluster transport config: %w", err)
+	}
+	kubeTransportConfig.TLS.NextProtos = []string{"http/1.1"}
+
+	kubeRoundTripper, err := transport.New(kubeTransportConfig)
+	if err != nil {
+		return nil, fmt.Errorf("could not get in-cluster transport: %w", err)
+	}
+
+	reverseProxy := httputil.NewSingleHostReverseProxy(serverURL)
+	reverseProxy.Transport = kubeRoundTripper
+
+	return &proxy{
+		cache: cache,
+		proxy: reverseProxy,
+		log:   log,
+	}, nil
+}
+
+func (p *proxy) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	log := p.log.WithValues(
+		"url", r.URL.String(),
+		"method", r.Method,
+	)
+
+	tokenCredentialReq, err := extractToken(r)
+	if err != nil {
+		log.Error(err, "invalid token encoding")
+		http.Error(w, "invalid token encoding", http.StatusBadRequest)
+		return
+	}
+	log = log.WithValues(
+		"authenticator", tokenCredentialReq.Spec.Authenticator,
+		"authenticatorNamespace", tokenCredentialReq.Namespace,
+	)
+
+	userInfo, err := p.cache.AuthenticateTokenCredentialRequest(r.Context(), tokenCredentialReq)
+	if err != nil {
+		log.Error(err, "received invalid token")
+		http.Error(w, "invalid token", http.StatusUnauthorized)
+		return
+	}
+
+	if userInfo == nil {
+		log.Info("received token that did not authenticate")
+		http.Error(w, "not authenticated", http.StatusUnauthorized)
+		return
+	}
+	log = log.WithValues("userID", userInfo.GetUID())
+
+	// Never mutate request (see http.Handler docs).
+	newR := r.WithContext(r.Context())
+	newR.Header = getProxyHeaders(userInfo, r.Header)
+
+	log.Info("proxying authenticated request")
+	p.proxy.ServeHTTP(w, newR)
+}
+
+func getProxyHeaders(userInfo user.Info, requestHeaders http.Header) http.Header {
+	newHeaders := http.Header{}
+	newHeaders.Set("Impersonate-User", userInfo.GetName())
+	for _, group := range userInfo.GetGroups() {
```

Comment:
We should be setting all the impersonation headers.  You might be able to use the impersonating round tripper from client-go.

Reply:
https://github.com/vmware-tanzu/pinniped/pull/355/commits/eb199801107cc55450fd982f4842e4d69b264a16